<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [39]:
import pandas as pd
import psycopg2

In [40]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 


In [41]:
connection = psycopg2.connect(
    dbname='project_sql',
    user='skillfactory',
    host='84.201.134.129',
    password='cCkxxLVrDE8EbvjueeMedPKt',
    port=5432
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [42]:

query_3_1 = f'''select id
           from public.vacancies
           count (id)
''' 

In [43]:

df = pd.read_sql_query(query_3_1, connection)
df

C:\Users\MAN-MADE\AppData\Local\Temp\ipykernel_9368\534329379.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


id
0      55312386
1      55843014
2      54525964
3      54525965
4      55354053
...         ...
49192  55589514
49193  55029668
49194  66172263
49195  66180086
49196  55547842

[49197 rows x 1 columns]

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [44]:
# выводим столбец id из таблицы работодателей и считаем их общее количество с помощью count 

query_3_2 = f'''select id
           from public.employers
           count (id)
''' 

In [45]:
# выполняем запрос

df = pd.read_sql_query(query_3_2, connection)
df

C:\Users\MAN-MADE\AppData\Local\Temp\ipykernel_9368\2881470143.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


id
0         2393
1        72977
2         3155
3          675
4      1840010
...        ...
23496  1441774
23497  4718575
23498  3702770
23499  5636092
23500  2064382

[23501 rows x 1 columns]

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [46]:
#  выводим столбец id из таблицы регионов вакансий и считаем их общее количество с помощью count 

query_3_3 = f'''select id
           from public.areas
           count (id)
''' 

In [47]:
# выполняем запрос

df = pd.read_sql_query(query_3_3, connection)
df

C:\Users\MAN-MADE\AppData\Local\Temp\ipykernel_9368\2059492830.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


id
0     2758
1        8
2     1180
3     1240
4        2
...    ...
1357  5043
1358  3934
1359  4825
1360  3708
1361  2432

[1362 rows x 1 columns]

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [48]:
# текст запроса

query_3_4 = f'''select id
           from public.industries
           count (id)
''' 

In [49]:
# результат запроса

df = pd.read_sql_query(query_3_4, connection)
df

C:\Users\MAN-MADE\AppData\Local\Temp\ipykernel_9368\204727846.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


id
0      7.540
1      7.539
2     27.550
3     27.551
4     13.664
..       ...
289   37.626
290   42.603
291   29.532
292   37.627
293  388.505

[294 rows x 1 columns]

***

__Выводы по предварительному анализу данных__
1. Исходя из предварительного анализа, можно сказать, что в базе данных присутствуют не только вакансии российских компаний,но и международных.
2. В среднем на одного работодателя приходится примерно две вакансии. 
3. Вакансии в базе данных представлены из широкого спектра индустрий, что говорит о высоком спросе на IT-специалистов на рынке труда.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [50]:
# объединяем таблицу вакансий с таблицей регионов по ключевому столбцу id региона, группируем id региона, выводим столбцы названия региона и кол-ва вакансий в нем

query_4_1 = f'''
select 
    a.name area,
    count (*) cnt
from vacancies v join areas a on v.area_id = a.id
group by a.id
order by cnt desc

''' 

In [51]:
# выполняем запрос

df = pd.read_sql_query(query_4_1, connection)
df

C:\Users\MAN-MADE\AppData\Local\Temp\ipykernel_9368\1242013880.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


area   cnt
0                      Москва  5333
1             Санкт-Петербург  2851
2                       Минск  2112
3                 Новосибирск  2006
4                      Алматы  1892
..                        ...   ...
764                    Кизляр     1
765                    Джизак     1
766                    Эртиль     1
767                  Арсеньев     1
768  Тура (Красноярский край)     1

[769 rows x 2 columns]

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [52]:
# текст запроса

query_4_2 = f'''
select 
    count(id) id_with_salary
from vacancies v 
where salary_from is not null or salary_to is not null

''' 

In [53]:
# результат запроса

df = pd.read_sql_query(query_4_2, connection)
df

C:\Users\MAN-MADE\AppData\Local\Temp\ipykernel_9368\1585536975.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


id_with_salary
0           24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [54]:
# текст запроса

query_4_3 = f'''
select 
    avg(salary_from) avg_from,
    avg(salary_to) avg_to
from vacancies v  

''' 

In [55]:
# результат запроса

df = pd.read_sql_query(query_4_3, connection)
df

C:\Users\MAN-MADE\AppData\Local\Temp\ipykernel_9368\3623202172.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


avg_from         avg_to
0  71064.657901  110536.741923

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [56]:
# текст запроса

query_4_4 = f'''
select 
    count(v.id),
    schedule,
    employment
from vacancies v
group by schedule, employment
order by 1 desc  

''' 

In [57]:
# результат запроса

df = pd.read_sql_query(query_4_4, connection)
df

C:\Users\MAN-MADE\AppData\Local\Temp\ipykernel_9368\64905383.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


count          schedule           employment
0   35367       Полный день     Полная занятость
1    7802  Удаленная работа     Полная занятость
2    1593     Гибкий график     Полная занятость
3    1312  Удаленная работа  Частичная занятость
4     940    Сменный график     Полная занятость
5     569       Полный день           Стажировка
6     367    Вахтовый метод     Полная занятость
7     347       Полный день  Частичная занятость
8     312     Гибкий график  Частичная занятость
9     141       Полный день     Проектная работа
10    133  Удаленная работа     Проектная работа
11    116     Гибкий график           Стажировка
12    101    Сменный график  Частичная занятость
13     64  Удаленная работа           Стажировка
14     18     Гибкий график     Проектная работа
15     12    Сменный график           Стажировка
16      2    Вахтовый метод     Проектная работа
17      1    Сменный график     Проектная работа

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [58]:
# текст запроса

query_4_5 = f'''
select 
    count(v.id),
    v.experience exp
from 
    vacancies v
group by v.experience
order by 1

''' 

In [59]:
# результат запроса

df = pd.read_sql_query(query_4_5, connection)
df

C:\Users\MAN-MADE\AppData\Local\Temp\ipykernel_9368\718884012.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


count                 exp
0   1337         Более 6 лет
1   7197           Нет опыта
2  14511       От 3 до 6 лет
3  26152  От 1 года до 3 лет

***

__Выводы по детальному анализу вакансий__

1. Из вышепроведенного анализа видно, что наибольшее количество вакансий расположено в крупных городах-миллионниках России и стран СНГ.
2. Большинство компаний интересует полная занятость сотрудника, но в то же время для крупной доли вакансий возможна удаленный вариант работы.
3. Средняя вилка зарплат среди рассмотренных вакансий составляет от 70 до 110 тысяч рублей, а значит, среднюю зарплату по все базе ваканский можно расценить около 90 тысяч рублей.
4. Компании чаще ищут сотрудников с начальным и средним опытом, это может говорить нам о том, что на рынке широко востребованы специалисты junior и middle сегмента


# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [60]:
# текст запроса
query_5_1 = f'''
select 
    e.name,
    count(v.id) cnt
from 
    public.vacancies v 
    join public.employers e on v.employer_id = e.id
group by e.name 
order by cnt desc
limit 5

''' 

In [61]:
# результат запроса
df = pd.read_sql_query(query_5_1, connection)
df

C:\Users\MAN-MADE\AppData\Local\Temp\ipykernel_9368\3834399874.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


name   cnt
0         Яндекс  1933
1     Ростелеком   491
2       Тинькофф   444
3           СБЕР   428
4  Газпром нефть   331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [62]:
# текст запроса
query_5_2 = f'''
select 
    a.name as "регион",
    count(distinct e.id) as "количество работодателей",
    count(distinct v.id) as "количество вакансий"
from 
    public.areas a
    left join public.employers e on a.id = e.area
    left join public.vacancies v on a.id = v.area_id
where v.id is null
group by a.id
order by "количество работодателей" desc
limit 1

''' 

In [63]:
# результат запроса

df = pd.read_sql_query(query_5_2, connection)
df

C:\Users\MAN-MADE\AppData\Local\Temp\ipykernel_9368\1941881734.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


регион  количество работодателей  количество вакансий
0  Россия                       410                    0

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [64]:
# текст запроса

query_5_3 = f'''
select 
    e.name as "работодатель",
    count(distinct v.area_id) as "количество регионов"
from 
    public.employers e
    left join public.vacancies v on e.id = v.employer_id
group by e.name
order by "количество регионов" desc


''' 

In [65]:
# результат запроса

df = pd.read_sql_query(query_5_3, connection)
df

C:\Users\MAN-MADE\AppData\Local\Temp\ipykernel_9368\3847170148.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


работодатель  количество регионов
0                      Яндекс                  181
1                  Ростелеком                  152
2                  Спецремонт                  116
3      Поляков Денис Иванович                   88
4                   ООО ЕФИН                    71
...                       ...                  ...
23170                  СДЕЛКА                    0
23171            Alandr Group                    0
23172                     СДК                    0
23173      Lemon Land Lombard                    0
23174                  Махина                    0

[23175 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [66]:
# текст запроса

query_5_4 = f'''
select 
    count(e.id) as "кол-во работ-лей без сферы"
from 
    public.employers e 
    left join employers_industries ei on ei.employer_id = e.id
where industry_id is null

''' 

In [67]:
# результат запроса

df = pd.read_sql_query(query_5_4, connection)
df

C:\Users\MAN-MADE\AppData\Local\Temp\ipykernel_9368\2092942460.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


кол-во работ-лей без сферы
0                        8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [68]:
# текст запроса

query_5_5 = f'''
select 
    e.name as "название компании"
from 
    public.employers e 
    left join employers_industries ei on ei.employer_id = e.id
group by e.name
HAVING count(ei.industry_id) = 4 
order by e.name
offset 2 limit 1

''' 

In [69]:
# результат запроса

df = pd.read_sql_query(query_5_5, connection)
df

C:\Users\MAN-MADE\AppData\Local\Temp\ipykernel_9368\3387268485.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


название компании
0              2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [70]:
# текст запроса

query_5_6 = f'''
select 
    count (id) as "количество работодателей"
from 
    public.employers_industries ei 
    left join industries i on ei.industry_id = i.id
where i.name = 'Разработка программного обеспечения'

''' 

In [71]:
# результат запроса

df = pd.read_sql_query(query_5_6, connection)
df

C:\Users\MAN-MADE\AppData\Local\Temp\ipykernel_9368\213170975.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


количество работодателей
0                      3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [72]:
# код для получения списка городов-милионников

import requests
import urllib.parse as req

path_common = req.urlparse('https://ru.wikipedia.org/wiki/Города-миллионеры_России#Список_городов-миллионеров')
url = path_common.scheme + '://' + req.quote(path_common.netloc) +  req.quote(path_common.path) + '#'+ req.quote(path_common.fragment)
df = pd.read_html(url)[1]
df

ImportError: lxml not found, please install it

In [ ]:
# текст запроса

In [ ]:
# результат запроса

__Выводы по анализу работодателей__
1. Из анализа работодателей заметно, что самым активным поиском сотрудников занимаются крупные российские компании (Сбер, Яндекс, Ростелеком)
2. Наиболее широко поиск новых специалистов представлен у компании Яндекс, в 181 регионе, что говорит нам о том, что компании ищет сотрудников не только России, но и за ее пределами.
3. Чаще всего компании интересуются разработчиками различного програмного обеспечения 

***

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [ ]:
# текст запроса

query_6_1 = f'''
select 
    count (*) as "data вакансии"
from 
    public.vacancies v
where v.name ILIKE '%data%' OR v.name ILIKE '%данн%'

''' 

In [ ]:
# результат запроса

df = pd.read_sql_query(query_6_1, connection)
df

C:\Users\MAN-MADE\AppData\Local\Temp\ipykernel_14524\2275399668.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


data вакансии
0           1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [ ]:
# текст запроса

query_6_2 = f'''
SELECT 
    COUNT(*) 
    
FROM public.vacancies v 

WHERE
    (v.name ILIKE '%data scientist%' OR
    v.name ILIKE '%data science%' OR
    v.name ILIKE '%исследователь данных%' OR
    v.name ILIKE '%machine learning%' OR
    (v.name ILIKE '%ML%' AND (v.name NOT ILIKE '%HTML%')) OR
    v.name ILIKE '%машинн%обучен%') AND
    (v.name ILIKE '%junior%' OR
    v.experience ILIKE '%нет опыта%' OR
    v.employment ILIKE '%стажировка%')

''' 

In [ ]:
# результат запроса

df = pd.read_sql_query(query_6_2, connection)
df

C:\Users\MAN-MADE\AppData\Local\Temp\ipykernel_14524\380269403.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [ ]:
# текст запроса

query_6_3 = f'''
SELECT 
    COUNT(*) 
    
FROM public.vacancies v 

WHERE
    (v.name ILIKE '%data scientist%' OR
    v.name ILIKE '%data science%' OR
    v.name ILIKE '%исследователь данных%' OR
    v.name ILIKE '%machine learning%' OR
    (v.name LIKE '%ML%' AND (v.name NOT ILIKE '%HTML%')) OR
    v.name ILIKE '%машинн%обучен%') AND 
    (v.key_skills ILIKE '%SQL%' OR
    v.key_skills ILIKE '%postgres%')
    

''' 

In [ ]:
# результат запроса

df = pd.read_sql_query(query_6_3, connection)
df

C:\Users\MAN-MADE\AppData\Local\Temp\ipykernel_14524\66396245.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


count
0    201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [ ]:
# текст запроса

query_6_4 = f'''
SELECT 
    COUNT(*) 
    
FROM public.vacancies v 

WHERE
    (v.name ILIKE '%data scientist%' OR
    v.name ILIKE '%data science%' OR
    v.name ILIKE '%исследователь данных%' OR
    v.name ILIKE '%machine learning%' OR
    (v.name LIKE '%ML%' AND (v.name NOT ILIKE '%HTML%')) OR
    v.name ILIKE '%машинн%обучен%') AND 
    (v.key_skills ILIKE '%Python%')
    

''' 

In [ ]:
# результат запроса

df = pd.read_sql_query(query_6_4, connection)
df

C:\Users\MAN-MADE\AppData\Local\Temp\ipykernel_14524\2776760131.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


count
0    351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [ ]:
# текст запроса

query_6_5 = f'''
SELECT 
    ROUND(AVG(length(key_skills) - length(REPLACE(key_skills, CHR(9), '' )))::numeric, 2)+1 
FROM public.vacancies v 
WHERE
    ((name ILIKE '%data scientist%') OR 
    (name ILIKE '%data science%') OR 
    (name ILIKE '%исследователь данных%') OR 
    ((name LIKE '%ML%') AND (name NOT ILIKE '%HTML%')) 
    OR (name ILIKE '%machine learning%') OR 
    (name ILIKE '%машинн%обучен%') 
    AND (key_skills IS NOT NULL))

''' 

In [ ]:
# результат запроса

df = pd.read_sql_query(query_6_5, connection)
df

C:\Users\MAN-MADE\AppData\Local\Temp\ipykernel_8484\3250313318.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


?column?
0      6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [ ]:
# текст запроса

query_6_6 = f'''
SELECT
    experience, 
    ROUND(AVG(COALESCE(((salary_from + salary_to)/2), salary_from, salary_to))) average_salary
FROM public.vacancies
WHERE ((name ILIKE '%data scientist%') 
    OR (name ILIKE '%data science%')  
    OR (name ILIKE '%исследователь данных%')    
    OR ((name LIKE '%ML%') 
    AND (name NOT ILIKE '%HTML%')) 
    OR (name ILIKE '%machine learning%') 
    OR (name ILIKE '%машинн%обучен%')) 
    AND ((salary_from IS NOT NULL) OR (salary_to IS NOT NULL))
GROUP BY experience

''' 



In [ ]:
# результат запроса

df = pd.read_sql_query(query_6_6, connection)
df

C:\Users\MAN-MADE\AppData\Local\Temp\ipykernel_8484\1626560934.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


experience  average_salary
0           Нет опыта         74643.0
1  От 1 года до 3 лет        139675.0
2       От 3 до 6 лет        243115.0

***

__Выводы по предметному анализу__

1. В целом рынок вакансий для дата-сайентистов широкий, но доля ваканский для новичков небольшая, это говорит, что работодателям зачастую требуются специалисты с опытом 
2. Для большой доли вакансий дата-сайентистов необходимы навыки python и postgres sql 
3. В среднем дата-сайентист с увеличением опыта на 3 года может претендовать на повышение средней зарплата примерно в два раза

# Общий вывод по проекту

__Итоги исследования__

Из проведенного исследования можно сделать несколько общих выводов:
- большое количество вакансии из базы данных относится к крупным городам, так как там чаще расположены крупные компании, нуждающиеся в IT-специалистах
- средняя зарплата среди всех вакансий около 90 тысяч рублей 
- компаниям нужны специалисты с набором из нескольких навыков, а так же приветствуется наличие опыта